cmd/web_https.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package cmd
+
+import (
+	"crypto/tls"
+	"net/http"
+	"os"
+	"strings"
+
+	"code.gitea.io/gitea/modules/graceful"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+	"github.com/klauspost/cpuid/v2"
+)
+
+var tlsVersionStringMap = map[string]uint16{
+	"":        tls.VersionTLS12, // Default to tls.VersionTLS12
+	"tlsv1.0": tls.VersionTLS10,
```

Comment:
We should not allow v1.0 and v1.1 since they are not secure any more.

Reply:
In real world, some users still need these legacy protocols.

In my company, one of our customers are still using TLSv1.1   😭

So, I think we just provide the options, whatever is used is the end user's choice.

Reply:
>  whatever is used is the end user's choice.

Should we at least advise on not recommending using the TLSv1.0 and TLSv1.1? Having as a option is fine, but it shouldn't be prominent that a user could think that's a good choice to choose it if they don't have a specific reason for it.

Reply:
Sure, we can write something about "not recommended to use xxx"

However, the default values of setting options are generally fine tuned, so if a user have to edit the setting to change the TLS protocols, they must be familiar the TLS system already and know what they are doing, our warning doesn't seem to help much, because everyone on internet will tell you that "only TLS>=1.2 is secure"

Reply:
> our warning doesn't seem to help much, because everyone on internet will tell you that "only TLS>=1.2 is secure"

At least we did warn, better safe than sorry.

Reply:
I guess a good example would be to look at the documentation for nginx, Apache and Tomcat

nginx's documentation:

http://nginx.org/en/docs/http/ngx_http_ssl_module.html#ssl_protocols

Apache's documentation:

https://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslprotocol

Tomcat appears to be completely silent too.

--- 

The point of this PR is to make Gitea configurable and provide the guns for people to shoot themselves should they want to do so or rather need to do so. It also updates our defaults to match a better standard. 

It very deliberately does not provide security advice. We do not have the expertise, time and patience to keep this up-to-date and if we start doing so we're absolutely going to have to keep it up-to-date. 

Are we really going to get into the habit of providing this level of security advice in Gitea? Do you want to be responsible for keeping this advice up-to-date and dealing with the negative comments from people who determine that we've done this wrong or not to their standards?

Feel free to write the documentation for this and add the warnings you feel necessary but honestly I think we would be better to be silent.

Reply:
Well - then I will say the magic words.
"Let it be", I do personally have a hard time with providing options to the user that seriously weaken the security. I do get that their are legitimate reasons for using TLSv1.0 and TLSv1.1 -  Given that the option is only be done after you explicitly typed the version, I think that should be fine (but it's for me personally on the edge).

The security advice that can be given possibly is widely accepted within the security community - In theoretical sense something like this could be given: "Changing this setting can have impact on the security of gitea, please read about what the TLS versions means over at: link". Making the user aware shouldn't do any harm(while like you say, people can and will disagree).

If in the future such issue or discussion come up, feel free, better I will insist that someone ping's me. I gladly will take part into such discussion. 

The configuring of the SSL chipher suite is good - The question if we should warn users can _possibly_ be done on another PR. Let's get this PR rolled into master.